src/charm.py

```diff
@@ -0,0 +1,465 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This connects the benchmark service to the database and the grafana agent.
+
+The first action after installing the benchmark charm and relating it to the different
+apps, is to prepare the db. The user must run the prepare action to create the database.
+
+The prepare action will run the benchmark prepare command to create the database and, at its
+end, it sets a systemd target informing the service is ready.
+
+The next step is to execute the run action. This action renders the systemd service file and
+starts the service. If the target is missing, then service errors and returns an error to
+the user.
+"""
+
+import logging
+import os
+from functools import cached_property
+from typing import Any, Optional
+
+import charms.operator_libs_linux.v0.apt as apt
+import ops
+from charms.data_platform_libs.v0.data_interfaces import KafkaRequires
+from charms.kafka.v0.client import KafkaClient, NewTopic
+from ops.charm import CharmBase
+from ops.framework import EventBase
+from ops.model import Application, BlockedStatus, Relation, Unit
+from overrides import override
+
+from benchmark.base_charm import DPBenchmarkCharmBase
+from benchmark.core.models import (
+    DatabaseState,
+    DPBenchmarkBaseDatabaseModel,
+    RelationState,
+)
+from benchmark.core.workload_base import WorkloadBase
+from benchmark.events.db import DatabaseRelationHandler
+from benchmark.events.peer import PeerRelationHandler
+from benchmark.literals import (
+    PEER_RELATION,
+    DPBenchmarkLifecycleTransition,
+)
+from benchmark.managers.config import ConfigManager
+from benchmark.managers.lifecycle import LifecycleManager
+from literals import CLIENT_RELATION_NAME, TOPIC_NAME
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+KAFKA_WORKER_PARAMS_TEMPLATE = """name: Kafka-benchmark
```

Comment:
**todo** this is waaay too low level for the `charm.py` module. This would seems appropriate in a manager actually